File: py/any2webp.py
# read any image and save webp, and keep exif

import sys

from PIL import Image
from pillow_heif import register_heif_opener

register_heif_opener()


def transform(input, output, max_height=1080):
    im = Image.open(input)

    # keep exif
    exif = im.info.get('exif')


    # reshape if needed
    if im.height > max_height:
        im = im.resize((int(im.width * max_height / im.height), max_height))

    if not output:
        output = input[:input.rfind('.')] + '.webp'

    if exif:
        im.save(output, 'WEBP', exif=exif)
    else:
        im.save(output, 'WEBP')

def main():
    if len(sys.argv) < 2:
        print('Usage: python any2webp.py <input> [output]')
        sys.exit(1)

    if len(sys.argv) < 3:
        input, output = sys.argv[1], ''
    else:
        input, output = sys.argv[1], sys.argv[2]

    if len(sys.argv) >= 4:
        max_height = int(sys.argv[3])
        transform(input, output, max_height)
    else:
        transform(input, output)


if __name__ == '__main__':
    main()

File: scripts/any2webp.js

import sharp from 'sharp';

async function transform(input, output, maxWidth = 1920) {
    let image = sharp(input);

    // Get metadata to check width and extract exif
    const metadata = await image.metadata();
    const exif = metadata.exif;

    // Resize if needed
    if (metadata.width > maxWidth) {
        image = image.resize(maxWidth);
    }

    if (!output) {
        output = input.slice(0, input.lastIndexOf('.')) + '.webp';
    }

    // Save image as webp with exif
    await image.webp({ quality: 90 }).withMetadata({ exif }).toFile(output);
}

async function main() {
    const args = process.argv.slice(2);
    if (args.length < 1) {
        console.log('Usage: node any2webp.js <input> [output]');
        process.exit(1);
    }

    const input = args[0];
    const output = args.length > 1 ? args[1] : '';

    await transform(input, output);
}

main().catch(err => {
    console.error(err);
    process.exit(1);
});